src/main/drivers/pwm_output.c

```diff
@@ -373,59 +373,6 @@ uint32_t getDshotHz(motorPwmProtocolTypes_e pwmProtocolType)
         return MOTOR_DSHOT150_HZ;
     }
 }
-
-void pwmWriteDshotCommand(uint8_t index, uint8_t motorCount, uint8_t command)
```

Comment:
`pwmWriteDshotCommand` and `prepareDshotPacket` are in here because they are shared between `src/main/drivers/pwm_output_dshot.c` and `src/main/drivers/pwm_output_dshot_hal.c`. Having identical copies of the functions in both files will increase the code size and effort for maintenance, and make code prone to breaking if one copy is updated but not the other. These functions should stay in here. Also, `dshotCommandControl_t` should be moved into here, alongside with functions for the code that uses it, for the same reason.

Reply:
Would it increase code size?  I thought `pwm_output_dshot.c` and `pwm_output_dshot_hal.c` were an either/or.

For `dshotCommandControl_t`,  it is being used in `pwmCompleteDshotMotorUpdate` I'd rather not make extra accessor functions to pull in the needed variables.  That would add complexity and increase code size.

`prepareDshotPacket` could be moved back, although the FAST_CODE seems like it really should be in `pwm_output_dshot_hal.c` and not in shared code.

As someone that is newer to the code it is confusing to find dshot related functions not in the dshot c files.  Maybe a pwm_output_dshot_common.c is needed?

Reply:
> Would it increase code size? I thought pwm_output_dshot.c and pwm_output_dshot_hal.c were an either/or.

You're right, it would not affect the code size. But the arguments about making the code harder to maintain, and more prone to bugs being introduced still stand. In general 'code reuse' by copy paste is to be avoided whenever possible.

> For dshotCommandControl_t, it is being used in pwmCompleteDshotMotorUpdate I'd rather not make extra accessor functions to pull in the needed variables. That would add complexity and increase code size.

It seems to be that `dshotCommandControl_t` is only needed in two pretty much self contained blocks. Making these blocks into functions in `src/main/drivers/pwm_output.c` and calling them from `src/main/drivers/pwm_output_dshot.c` / `src/main/drivers/pwm_output_dshot_hal.c` will look tidy, and, due to the inlining that is done as part of the optimisation, will not result in undue overhead.

> As someone that is newer to the code it is confusing to find dshot related functions not in the dshot c files. Maybe a pwm_output_dshot_common.c is needed?

Yes, `pwm_output_dshot_common.c` is an option. Or `pwm_output.c` could be renamed to `pwm_output_common.c` to make this clearer.

Reply:
`FAST_CODE` is defined to emptiness on anything besides F7 (for now). If we had a function copy for each time we wanted to mark it `FAST_CODE`, it would complicate maintenance severely while giving nothing in return. If `prepareDshotPacket` is a duplicate, it should really be moved into the common file.
The issue is more with adding new logic into `pwmWriteDshotInt`. I carefully profiled and measured it when optimizing F7, and adding new code into it is not an option if we want to retain F7 performance.